### PR TITLE
Implement DictSquashUpdatePtr

### LINKED ISSUE #293: DictSquashUpdatePtr
Updates the pointer of the squashed dictionary to point to the end of the new dictionary.

Reference: https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/cairo/common/dict.cairo#L106

### PR DESCRIPTION
**Pull request by har777**
Fixes #293

### CODE REVIEW

File: pkg/hintrunner/hinter/zero_dict.go [hunk 1]
```diff
@@ -39,6 +39,11 @@ func (d *ZeroDictionary) IncrementFreeOffset(freeOffset uint64) {
 	d.freeOffset += freeOffset
 }
 
+// Given a freeOffset value, it sets the freeOffset field of dictionary by it
```
Comment (TAdev0) [inline pkg/hintrunner/hinter/zero_dict.go:42]: > ... by it


to it?

File: pkg/hintrunner/hinter/zero_dict.go [hunk 2]
```diff
@@ -107,9 +112,18 @@ func (dm *ZeroDictionaryManager) Set(dictAddr mem.MemoryAddress, key f.Element,
...
+}
+
+// Given a memory address and a freeOffset, it sets the freeOffset field of dictionary to it.
+func (dm *ZeroDictionaryManager) SetFreeOffset(dictAddr mem.MemoryAddress, freeOffset uint64) error {
 	if dict, ok := dm.dictionaries[dictAddr.SegmentIndex]; ok {
-		dict.IncrementFreeOffset(freeOffset)
+		dict.SetFreeOffset(freeOffset)
 		return nil
 	}
 	return fmt.Errorf("no dictionary at address: %s", dictAddr)
```
Comment (TAdev0) [inline pkg/hintrunner/hinter/zero_dict.go:129]: nice

File: pkg/hintrunner/zero/hintcode.go [hunk 3]
```diff
@@ -121,6 +121,7 @@ const (
 	squashDictInnerLenAssert          string = "assert len(current_access_indices) == 0"
 	squashDictInnerNextKey            string = "assert len(keys) > 0, 'No keys left but remaining_accesses > 0.'\nids.next_key = key = keys.pop()"
 	squashDictInnerUsedAccessesAssert string = "assert ids.n_used_accesses == len(access_indices[key])"
+	dictSquashUpdatePtrCode           string = "# Update the DictTracker's current_ptr to point to the end of the squashed dict.\n__dict_manager.get_tracker(ids.squashed_dict_start).current_ptr = \\\n    ids.squashed_dict_end.address_"
```
Comment (TAdev0) [inline pkg/hintrunner/zero/hintcode.go:124]: why did you add ''Code'' at the end of the hint name?

File: pkg/hintrunner/zero/zerohint.go [hunk 4]
```diff
@@ -181,6 +181,8 @@ func GetHintFromCode(program *zero.ZeroProgram, rawHint zero.Hint, hintPC uint64
 		return createSquashDictInnerNextKeyHinter(resolver)
 	case squashDictInnerUsedAccessesAssert:
 		return createSquashDictInnerUsedAccessesAssertHinter(resolver)
+	case dictSquashUpdatePtrCode:
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint.go:184]: same

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 5]
```diff
@@ -398,3 +398,48 @@ func createSquashDictInnerUsedAccessesAssertHinter(resolver hintReferenceResolve
 
 	return newSquashDictInnerUsedAccessesAssertHint(nUsedAccesses), nil
 }
+
+// DictSquashUpdatePtrCode updates the DictTracker's current_ptr to point to the end of the squashed dict
+//
+// `newDictSquashUpdatePtrCodeHint` takes two operanders as arguments
+//   - `squashed_dict_start` pointer to the dictionary whose current_ptr should be updated
+//   - `squashed_dict_end` new current_ptr of the dictionary
+func newDictSquashUpdatePtrCodeHint(squashedDictStart, squashedDictEnd hinter.ResOperander) hinter.Hinter {
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:407]: same , there is Code everywhere, i think it can be removed?

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 6]
```diff
@@ -398,3 +398,48 @@ func createSquashDictInnerUsedAccessesAssertHinter(resolver hintReferenceResolve
...
+			if err != nil {
+				return err
+			}
+
+			dictionaryManager, ok := ctx.ScopeManager.GetZeroDictionaryManager()
+			if !ok {
+				return fmt.Errorf("__dict_manager not in scope")
+			}
+
+			//> dict_tracker.current_ptr += ids.DictAccess.SIZE
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:427]: what is this comment? its not part of the hint is it?

File: pkg/hintrunner/zero/hintcode.go (hunk 3 above)
Comment (har777) [reply pkg/hintrunner/zero/hintcode.go:124]: Do you mean at the end of the variable name? All variables containing the hint string should end with `Code` right 🤔

Comment (TAdev0) [reply pkg/hintrunner/zero/hintcode.go:124]: oh yeah indeed! well we forgot the ''Code''at the end for a few hintcode string in hintcode.go , mostly mine actually ahha
will solve that in another PR

**Review by TAdev0 — APPROVED** [review]
lgtm!

**Comment by TAdev0** [comment]
@har777  you have a conflict, can you resolve it?

**Comment by har777** [comment]
> @har777 you have a conflict, can you resolve it?

Done!

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 7]
```diff
@@ -722,3 +722,47 @@ func createSquashDictInnerUsedAccessesAssertHinter(resolver hintReferenceResolve
...
+			if err != nil {
+				return err
+			}
+
+			dictionaryManager, ok := ctx.ScopeManager.GetZeroDictionaryManager()
+			if !ok {
+				return fmt.Errorf("__dict_manager not in scope")
+			}
+
+			// TODO: figure out if its ever possible for squashedDictEnd segment to be different from the dictionary segment
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:751]: I think it is possible, however it is going to be some weird edge case or intentional endurance test for the VM. Maybe we could try to include that in the integration tests?

**Review by MaksymMalicki — APPROVED** [review]

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 7 above)
Comment (har777) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:751]: These hints are only used by the dict library which wont have such a scenario. But we can manually write code which calls the hints with such a scenario. Do you think maybe I should just check if  `squashedDictStart` and `squashedDictEnd` are of the same segment before setting the offset? Just to be safe 🤔

Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:846]: my opinion is that it seems useless. i dont know. If the VM runs normal cairo program, i dont see any situation where this could happen mmh..

Comment (cicr99) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:846]: Yeah, I agree with Tristan, don't think there is a need for us to test this if it's not a realistic scenario

File: pkg/hintrunner/hinter/zero_dict.go [hunk 8]
```diff
@@ -39,6 +39,11 @@ func (d *ZeroDictionary) IncrementFreeOffset(freeOffset uint64) {
 	*d.FreeOffset += freeOffset
 }
 
+// Given a freeOffset value, it sets the freeOffset field of dictionary to it
+func (d *ZeroDictionary) SetFreeOffset(freeOffset uint64) {
+	*d.FreeOffset = freeOffset
+}
```
[the remark this answers is not in the document]
Comment (har777) [reply pkg/hintrunner/hinter/zero_dict.go:45]: Thats a good idea. Actually I'm going to make all dict functions private. I think its better to interact with dicts via its manager!
